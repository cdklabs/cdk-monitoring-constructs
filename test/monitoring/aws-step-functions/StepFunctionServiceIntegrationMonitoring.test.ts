import { Duration, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import {
  AlarmWithAnnotation,
  StepFunctionServiceIntegrationMonitoring,
} from "../../../lib";
import { TestMonitoringScope } from "../TestMonitoringScope";

test("snapshot test: no alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  new StepFunctionServiceIntegrationMonitoring(scope, {
    alarmFriendlyName: "DummyServiceIntegration",
    serviceIntegrationResourceArn:
      "arn:aws:states:eu-west-1:216854468193:batch:submitJob.sync",
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("snapshot test: all alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  let numAlarmsCreated = 0;

  new StepFunctionServiceIntegrationMonitoring(scope, {
    alarmFriendlyName: "DummyServiceIntegration",
    serviceIntegrationResourceArn:
      "arn:aws:states:eu-west-1:216854468193:batch:submitJob.sync",
    addDurationP50Alarm: {
      Warning: {
        maxDuration: Duration.minutes(1),
      },
    },
    addDurationP90Alarm: {
      Warning: {
        maxDuration: Duration.minutes(2),
      },
    },
    addDurationP99Alarm: {
      Warning: {
        maxDuration: Duration.minutes(3),
      },
    },
    addFailedServiceIntegrationsCountAlarm: {
      Warning: {
        maxErrorCount: 2,
      },
    },
    addFailedServiceIntegrationsRateAlarm: {
      Warning: {
        maxErrorRate: 1,
      },
    },
    addTimedOutServiceIntegrationsCountAlarm: {
      Warning: {
        maxErrorCount: 2,
      },
    },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreated = alarms.length;
      },
    },
  });

  expect(numAlarmsCreated).toStrictEqual(6);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
