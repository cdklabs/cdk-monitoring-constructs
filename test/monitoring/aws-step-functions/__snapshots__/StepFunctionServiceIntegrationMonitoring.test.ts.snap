// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: all alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ScopeTestDummyServiceIntegrationDurationP50Warning9B6E5107": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P50 duration is too long.",
        "AlarmName": "Test-DummyServiceIntegration-Duration-P50-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P50",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ServiceIntegrationResourceArn",
                    "Value": "arn:aws:states:eu-west-1:216854468193:batch:submitJob.sync",
                  },
                ],
                "MetricName": "ServiceIntegrationRunTime",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "p50",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 60000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyServiceIntegrationDurationP90Warning7BF80296": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P90 duration is too long.",
        "AlarmName": "Test-DummyServiceIntegration-Duration-P90-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P90",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ServiceIntegrationResourceArn",
                    "Value": "arn:aws:states:eu-west-1:216854468193:batch:submitJob.sync",
                  },
                ],
                "MetricName": "ServiceIntegrationRunTime",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "p90",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 120000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyServiceIntegrationDurationP99Warning0386A528": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P99 duration is too long.",
        "AlarmName": "Test-DummyServiceIntegration-Duration-P99-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P99",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ServiceIntegrationResourceArn",
                    "Value": "arn:aws:states:eu-west-1:216854468193:batch:submitJob.sync",
                  },
                ],
                "MetricName": "ServiceIntegrationRunTime",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "p99",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 180000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyServiceIntegrationFailureCountWarningF0D0C79F": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Failure count is too high.",
        "AlarmName": "Test-DummyServiceIntegration-Failure-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Failed",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ServiceIntegrationResourceArn",
                    "Value": "arn:aws:states:eu-west-1:216854468193:batch:submitJob.sync",
                  },
                ],
                "MetricName": "ServiceIntegrationsFailed",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyServiceIntegrationFailureRateWarningFB59B204": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Failure rate is too high.",
        "AlarmName": "Test-DummyServiceIntegration-Failure-Rate-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Failed (avg)",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ServiceIntegrationResourceArn",
                    "Value": "arn:aws:states:eu-west-1:216854468193:batch:submitJob.sync",
                  },
                ],
                "MetricName": "ServiceIntegrationsFailed",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyServiceIntegrationTimedOutCountWarning84407F1A": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "TimedOut count is too high.",
        "AlarmName": "Test-DummyServiceIntegration-TimedOut-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Timeout",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ServiceIntegrationResourceArn",
                    "Value": "arn:aws:states:eu-west-1:216854468193:batch:submitJob.sync",
                  },
                ],
                "MetricName": "ServiceIntegrationsTimedOut",
                "Namespace": "AWS/States",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`snapshot test: no alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
